sdk/communication/AzureCommunicationChat/Source/Signaling/CommunicationSignalingClient.swift

```diff
@@ -0,0 +1,179 @@
+    ) throws {
+        self.communicationSkypeTokenProvider = skypeTokenProvider
+        self.logger = logger
+
+        let trouterSkypeTokenHeaderProvider = TrouterSkypetokenAuthHeaderProvider(
+            skypetokenProvider: communicationSkypeTokenProvider
+        )
+
+        let communicationCache = CommunicationCache()
+        selfHostedTrouterClient = SelfHostedTrouterClient.create(
+            withClientVersion: defaultClientVersion,
+            authHeadersProvider: trouterSkypeTokenHeaderProvider,
+            dataCache: communicationCache,
+            trouterHostname: defaultTrouterHostname
+        )
+
+        guard let regData = defaultRegistrationData else {
+            throw AzureError.client("Failed to create TrouterUrlRegistrationData.")
+        }
+
+        guard let trouterUrlRegistrar = TrouterUrlRegistrar.create(
+            with: communicationSkypeTokenProvider,
+            registrationData: regData,
+            registrarHostnameAndBasePath: defaultRegistrarHostnameAndBasePath,
+            maxRegistrationTtlS: 3600
+        ) as? TrouterUrlRegistrar else {
+            throw AzureError.client("Failed to create TrouterUrlRegistrar.")
+        }
+
+        self.trouterUrlRegistrar = trouterUrlRegistrar
+    }
+
+    convenience init(
+        token: String
+    ) throws {
+        let skypeTokenProvider = CommunicationSkypeTokenProvider(skypeToken: token)
+        try self.init(skypeTokenProvider: skypeTokenProvider)
+    }
+
+    func start() {
+        selfHostedTrouterClient.withRegistrar(trouterUrlRegistrar)
+        selfHostedTrouterClient.start()
+    }
+
+    func stop() {
+        selfHostedTrouterClient.stop()
+        communicationHandlers.forEach { _, handler in
+            selfHostedTrouterClient.unregisterListener(handler)
+        }
+        communicationHandlers.removeAll()
+    }
+
+    func on(event: ChatEventId, handler: @escaping EventHandler) {
+        let communicationHandler = CommunicationHandler(handler: handler)
+        selfHostedTrouterClient.register(communicationHandler, forPath: "/\(event)")
+        communicationHandlers[event] = communicationHandler
+    }
+
+    func off(event: ChatEventId) {
+        if let communicationHandler = communicationHandlers[event] {
+            selfHostedTrouterClient.unregisterListener(communicationHandler)
+            communicationHandlers.removeValue(forKey: event)
+        }
+    }
+}
+
+class CommunicationSkypeTokenProvider: NSObject, TrouterSkypetokenProvider {
+    var skypeToken: String?
+    func getSkypetoken(_: Bool) -> String! {
+        return skypeToken
+    }
+
+    init(skypeToken: String? = nil) {
+        self.skypeToken = skypeToken
+    }
+}
+
+class CommunicationCache: NSObject, TrouterConnectionDataCache {
+    var data: String?
+
+    func store(_ data: String!) {
+        self.data = data
+    }
+
+    func load() -> String {
+        return data ?? ""
+    }
+}
```

Comment:
```
/**
 * Represents a callback for trouter client to store and load connection data.
 * Storing that data allows Trouter Client to reconnect to the service quicker
 * and preserves Trouter URL.
 *
 * It can even be a simple in-memory storage for applications that start/stop
 * Trouter client often during its lifetime
 *
 * !!! It is important to carefully separate data for applications with
 *  multiple Trouter connections (e.g. multi user login)
 */
TROUTERCLIENTIOS_EXPORT
@protocol TrouterConnectionDataCache <NSObject>
/**
 * Called when trouter establishes new connection or is forced to reconnect to a new DC
 */
-(void)store:(NSString*)data;
/**
 * Called when trouter starts up and checks if it can reconnect instead of
 * establishing brand new connection
 */
-(NSString*)load;
@end
```